### PR TITLE
8287661: Fix and improve BitMap::print_on(outputStream*)

### PR DESCRIPTION
**Pull request by shipilev**
Found these inconvenient during debugging using the `BitMap::print_on`. We pass the `outputStream*` to the method, but do not use it. Also, the output is hard to read for large bitmaps. 

Sample output generated by gtest:

```
[0.292s][info][test] Bitmap (0 bits):
[0.292s][info][test] Bitmap (128 bits):
[0.292s][info][test]     0: ........ ........ ........ ........ ........ ........ ........ ........
[0.292s][info][test]    64: S....... ........ ........ ........ ........ ........ ........ ........
[0.292s][info][test] Bitmap (1024 bits):
[0.292s][info][test]     0: ........ ........ ........ ........ ........ ........ ........ ........
[0.292s][info][test]    64: ........ ........ ........ ........ ........ ........ ........ ........
[0.292s][info][test]   128: ........ ........ ........ ........ ........ ........ ........ ........
[0.292s][info][test]   192: ........ ........ ........ ........ ........ ........ ........ ........
[0.292s][info][test]   256: ........ ........ ........ ........ ........ ........ ........ ........
[0.292s][info][test]   320: ........ ........ ........ ........ ........ ........ ........ ........
[0.292s][info][test]   384: ........ ........ ........ ........ ........ ........ ........ ........
[0.292s][info][test]   448: ........ ........ ........ ........ ........ ........ ........ ........
[0.292s][info][test]   512: S....... ........ ........ ........ ........ ........ ........ ........
[0.292s][info][test]   576: ........ ........ ........ ........ ........ ........ ........ ........
[0.292s][info][test]   640: ........ ........ ........ ........ ........ ........ ........ ........
[0.292s][info][test]   704: ........ ........ ........ ........ ........ ........ ........ ........
[0.292s][info][test]   768: ........ ........ ........ ........ ........ ........ ........ ........
[0.292s][info][test]   832: ........ ........ ........ ........ ........ ........ ........ ........
[0.292s][info][test]   896: ........ ........ ........ ........ ........ ........ ........ ........
[0.292s][info][test]   960: ........ ........ ........ ........ ........ ........ ........ ........
```

It becomes clearer that bits 64 and 512 are set, respectively.

Additional testing:
 - [x] Linux x86_64 fastdebug, new gtest
 - [x] Linux x86_64 release, new gtest

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.java.net/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8287661](https://bugs.openjdk.java.net/browse/JDK-8287661): Fix and improve BitMap::print_on(outputStream*)

### Reviewers
 * [Stefan Karlsson](https://openjdk.java.net/census#stefank) (@stefank - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/8970/head:pull/8970` \
`$ git checkout pull/8970`

Update a local copy of the PR: \
`$ git checkout pull/8970` \
`$ git pull https://git.openjdk.java.net/jdk pull/8970/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 8970`

View PR using the GUI difftool: \
`$ git pr show -t 8970`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/8970.diff">https://git.openjdk.java.net/jdk/pull/8970.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/utilities/bitMap.cpp [hunk 1]
```diff
@@ -691,11 +691,18 @@ void BitMap::write_to(bm_word_t* buffer, size_t buffer_size_in_bytes) const {
...
+      st->print(SIZE_FORMAT_W(5) ":", index);
+    }
+    if ((index % 10) == 0) {
+      st->print(" ");
+    }
+    st->print("%c", at(index) ? 'S' : '.');
   }
-  tty->cr();
+  st->cr();
 }
```
Comment (stefank) [inline src/hotspot/share/utilities/bitMap.cpp:706]: I wonder if it wouldn't be more useful to use line and chunk sizes that are a power-of-two? Maybe if we find that that would be more useful, than we could later add a second function to support that.

Comment (shipilev) [reply src/hotspot/share/utilities/bitMap.cpp:706]: My thought was that the *humans* would try to see where the set bits are, and so working in decimals is better? But we can do binary too, maybe grouping by 8 bits and 8 bytes per line would be quite natural.

Comment (shipilev) [reply src/hotspot/share/utilities/bitMap.cpp]: Done, see new commit, and new PR body for the example.

File: test/hotspot/gtest/utilities/test_bitMap.cpp [hunk 2]
```diff
@@ -95,6 +95,19 @@ class BitMapTest {
...
+#ifndef PRODUCT
+
+  static void testPrintOn(BitMap::idx_t size) {
+    ResourceMark rm;
+
+    ResourceBitMap map(size);
+    if (size > 0) {
+      map.set_bit(size / 2);
+    }
+    map.print_on(tty);
```
[the remark this answers is not in the document]
Comment (shipilev) [reply test/hotspot/gtest/utilities/test_bitMap.cpp]: Done.

**Review by stefank — APPROVED** [review]
Looks great! If we later change our mind about base 10 vs 2, we can revisit this.

**Comment by shipilev** [comment]
Any other opinions? I'll integrate it otherwise.

**Comment by DamonFool** [comment]
> Any other opinions? I'll integrate it otherwise.

There seems to be a regression: https://github.com/openjdk/jdk/pull/9029 .
Please have a look.
Thanks.

**Pull request closed without merging**